File: ruby/rna-transcription/rna_transcription.rb
class Complement
  DNA_TO_RNA = { "G" => "C", "C" => "G", "T" => "A", "A" => "U" }
  def self.of_dna(dna)
    return '' if dna =~ /[^GCTA]+/
    dna.gsub(/[GCTA]/, DNA_TO_RNA) 
  end
end

File: python/rna-transcription/rna_transcription.py
import re

DNA_TO_RNA_MAPPING = str.maketrans('GCTA', 'CGAU')

def to_rna(dna):
    """Returns the RNA complement of a given DNA strand.

    :param dna (str): the DNA strand.
    :returns: the RNA complement.
    :raises ValueError: if the argument dna contains characters other than
    [GCTA]
    """
    not_dna_chars = re.compile(r'[^GCTA]|^$')
    if not_dna_chars.search(dna):
        raise ValueError(f'cannot contain characters other than [GCTA]')
    return dna.translate(DNA_TO_RNA_MAPPING)
